src/main/java/com/google/cloud/tools/managedcloudsdk/internal/download/StandardDownloader.java

```diff
@@ -0,0 +1,100 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.managedcloudsdk.internal.download;
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.NotDirectoryException;
+import java.nio.file.Path;
+
+/** Standard implementation of {@link Downloader}. */
+public final class StandardDownloader implements Downloader {
+
+  static final int BUFFER_SIZE = 8 * 1024;
+  private final URL address;
+  private final Path destinationFile;
+  private final String userAgentString;
+  private final DownloadProgressListener downloadProgressListener;
+
+  /** Use {@link DownloaderFactory} to instantiate. */
+  StandardDownloader(
+      URL source,
+      Path destinationFile,
+      String userAgentString,
+      DownloadProgressListener downloadProgressListener) {
+    this.address = source;
+    this.destinationFile = destinationFile;
+    this.userAgentString = userAgentString;
+    this.downloadProgressListener = downloadProgressListener;
+  }
+
+  @Override
+  public Path call() throws IOException, InterruptedException {
+    Path destinationDir = destinationFile.getParent();
+    if (Files.exists(destinationDir) && !Files.isDirectory(destinationDir)) {
+      throw new NotDirectoryException(
+          "Cannot download to " + destinationDir + " because it is not a directory");
+    }
+    if (Files.exists(destinationFile)) {
+      throw new FileAlreadyExistsException(
+          "Cannot write to " + destinationFile + " because it already exists");
+    }
+    if (!Files.exists(destinationDir)) {
+      Files.createDirectories(destinationDir);
+    }
+
+    URLConnection connection = address.openConnection();
+    connection.setRequestProperty("User-Agent", userAgentString);
+
+    try (BufferedOutputStream out =
+        new BufferedOutputStream(Files.newOutputStream(destinationFile))) {
+      try (InputStream in = connection.getInputStream()) {
+
+        long contentLength = connection.getContentLengthLong();
+        // Progress is updated every 1%
+        long updateThreshold = contentLength / 100;
+
+        byte[] buffer = new byte[BUFFER_SIZE];
+        long lastUpdated = 0;
+        long totalBytesRead = 0;
+
+        int bytesRead;
+        while ((bytesRead = in.read(buffer)) != -1) {
+          if (Thread.currentThread().isInterrupted()) {
+            throw new InterruptedException("Downloader was interrupted.");
+          }
+          totalBytesRead += bytesRead;
```

Comment:
Since this loop does two things (writing the read bytes to the output file, and updating the progress bar), it might be good to separate out the progress update into another function to make this function easier to read, like move lines 74, 77-78, and 85-93 to a helper function like `updateProgress(bytesRead, contentLength)`, and maintain the state in member variables or a separate download state DDO.

Reply:
moved stuff around a little, adding a second function would probably require us to save state across calls, so I avoided that, hopefully it's easier to read now.